### PR TITLE
[NEW] Use REST API to upload files with fallback to older servers

### LINKED ISSUE #1055: [NEW] Support uploading a file using REST API
DOCS: https://github.com/RocketChat/Rocket.Chat/pull/8947

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

Closes #1055

### CODE REVIEW

**Comment by cardoso** [comment]
This is not working yet...

I'm receiving this error from the server:

```swift
{
    "error" : "File required", 
    "success" : false
}
```

For reference:
[API](https://github.com/RocketChat/Rocket.Chat/blob/5bd2cb3c6d408392891c07f26a24f5ea98421b25/packages/rocketchat-api/server/v1/rooms.js#L28)
[Busboy](https://github.com/mscdex/busboy/blob/master/lib/types/multipart.js#L174)

**Comment by rafaelks** [comment]
@cardoso Take a look here in the code below, we're using a different `Content-Type` for the file and the order of the parameters are also different.

https://github.com/RocketChat/Rocket.Chat.iOS/blob/ce206a6d036e9bda3170d958a87f492a7050ae23/Rocket.Chat/Managers/Uploader/UploadManager.swift#L42-L60

File: Rocket.Chat/Managers/Uploader/UploadManager.swift [hunk 1]
```diff
@@ -76,6 +76,20 @@ class UploadManager {
 
     // swiftlint:disable function_body_length
     func uploadToUFSFile(store: String, file: FileUpload, subscription: Subscription, progress: UploadProgressBlock, completion: @escaping UploadCompletionBlock) {
+
+        if let api = API.current() {
```
[the remark this answers is not in the document]
Comment (cardoso) [reply Rocket.Chat/Managers/Uploader/UploadManager.swift]: Not how it's working currently. I use the Auth object from AuthManager.isAuthenticated() to build it.

**Comment by cardoso** [comment]
@rafaelks ready for review 👍

File: Rocket.Chat/Managers/Uploader/UploadManager.swift [hunk 2]
```diff
@@ -76,76 +76,98 @@ class UploadManager {
...
-                        "method": "ufsComplete",
-                        "params": [fileIdentifier, normalizedStore, fileToken]
-                    ] as [String: Any]
-
-                    SocketManager.send(request) { [unowned self] (response) in
-                        guard !response.isError() else {
-                            completion(response, true)
-                            return
-                        }
+                    completion(nil, true)
```
Comment (rafaelks) [inline Rocket.Chat/Managers/Uploader/UploadManager.swift]: Can we send the error to the completion block? I think would be interesting to show the error message here, example: `file limit is 2MB`.

**Comment by cardoso** [comment]
Blocked by https://github.com/RocketChat/Rocket.Chat/issues/9096

**Comment by cardoso** [comment]
@rafaelks now showing errors (unblocked by https://github.com/RocketChat/Rocket.Chat/pull/9108)

File: Rocket.Chat/API/Clients/UploadClient.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                // TODO: Remove Upload fallback after Rocket.Chat 1.0
+                versionFallback?()
+            } else {
+                Alert(key: "alert.upload_error").present()
+                completion?()
+            }
+        })
+    }
+
+    func deletePushToken(token: String? = PushManager.getDeviceToken()) {
```
Comment (rafaelks) [inline Rocket.Chat/API/Clients/UploadClient.swift]: Why is deletePushToken method here? 🤔

File: Rocket.Chat/API/Clients/UploadClient.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        let req = UploadRequest(
+            roomId: roomId,
+            data: data,
+            filename: filename,
+            mimetype: mimetype
+        )
+
+        api.fetch(req, succeeded: { result in
+            if let error = result.error {
+                Alert(key: "alert.upload_error").withMessage(error).present()
```
Comment (rafaelks) [inline Rocket.Chat/API/Clients/UploadClient.swift]: status code 2xx with error (I know it's API's fault)? 😂

File: Rocket.Chat/API/Clients/UploadClient.swift (hunk 3 above)
Comment (cardoso) [reply Rocket.Chat/API/Clients/UploadClient.swift]: No good reason 😊

File: Rocket.Chat/API/Clients/UploadClient.swift (hunk 4 above)
Comment (cardoso) [reply Rocket.Chat/API/Clients/UploadClient.swift]: 🤷‍♀️

**Pull request merged**